pallets/xyk/src/lib.rs

```diff
@@ -569,39 +592,100 @@ pub mod pallet {
+
+	pub fn get_liquidity_minting_user(
+		user: AccountIdOf<T>,
 		liquidity_asset_id: TokenId,
-		block_number: u32,
-	) -> Result<Balance, DispatchError> {
-		let already_claimed_pool = LiquidityMiningPoolClaimed::<T>::try_get(&liquidity_asset_id)
-			.unwrap_or_else(|_| 0 as u128);
+		current_time: u32,
+	) -> Result<(u32, U256, U256), DispatchError> {
+		let liquidity_assets_amount =
+			<T as Config>::Currency::total_balance(liquidity_asset_id.into(), &user).into();
 
-		let time = block_number / TIMEBLOCKRATIO; // round to time metric, as other liq minting functions
-		let available_rewards_for_pool: Balance = u128::from(time)
-			.checked_mul(1000) // back to block number
-			.ok_or_else(|| DispatchError::from(Error::<T>::MathOverflow))?
-			.checked_mul(1000) // mga minted per block
-			.ok_or_else(|| DispatchError::from(Error::<T>::MathOverflow))?
-			.checked_sub(already_claimed_pool)
-			.ok_or_else(|| DispatchError::from(Error::<T>::NotEnoughtRewardsEarned))?;
+		// user already minted after pool promotion
+		if LiquidityMiningUser::<T>::contains_key((&user, &liquidity_asset_id)) {
+			Ok(LiquidityMiningUser::<T>::get((&user, &liquidity_asset_id)))
+		}
+		// users first mint
+		else if liquidity_assets_amount == Balance::try_from(0).unwrap() {
+			Ok((current_time, U256::from(0), U256::from(0)))
+		}
+		// user minted liquidity before pool promotion, need to calculate checkpoint from poolPromotionStart, till now, since there is no data stored
+		else {
+			let pool_promotion_start = PoolPromotionStart::<T>::get(liquidity_asset_id);
+
+			let user_work_total = Self::calculate_work(
+				liquidity_assets_amount,
+				current_time,
+				pool_promotion_start,
+				U256::from(0),
+				liquidity_assets_amount.into(),
+			)?;
+			let time_passed = current_time
+				.checked_sub(pool_promotion_start)
+				.ok_or_else(|| DispatchError::from(Error::<T>::PastTimeCalculation))?;
+			let user_missing_at_checkpoint = Self::calculate_missing_at_checkpoint(
+				time_passed,
+				Balance::try_from(0).unwrap(),
+				liquidity_assets_amount.into(),
+			)?;
 
-		Ok(available_rewards_for_pool)
+			LiquidityMiningUser::<T>::try_get((&user, &liquidity_asset_id))
+				.unwrap_or_else(|_| (0, U256::from(0), U256::from(0)));
+
+			Ok((current_time, user_work_total, user_missing_at_checkpoint))
+		}
+	}
+
+	pub fn calculate_rewards_amount(
+		user: AccountIdOf<T>,
+		liquidity_asset_id: TokenId,
+	) -> Result<(Balance, Balance), DispatchError> {
+		ensure!(
+			PoolPromotionStart::<T>::contains_key(liquidity_asset_id),
+			Error::<T>::NotAPromotedPool
+		);
+		let current_time: u32 =
+			<frame_system::Pallet<T>>::block_number().saturated_into::<u32>() / TIMEBLOCKRATIO;
+		let work_user = Self::calculate_work_user(user.clone(), liquidity_asset_id, current_time)?;
+		let work_pool = Self::calculate_work_pool(liquidity_asset_id, current_time)?;
+
+		let burned_not_claimed_rewards =
+			LiquidityMiningUserToBeClaimed::<T>::try_get((user, &liquidity_asset_id))
```

Comment:
use regular ::get instead, zeros will be returned as default

Reply:
replaced by get